nixos/modules/security/wrappers/wrapper.c

```diff
@@ -161,22 +162,25 @@ static int make_caps_ambient(const char *selfPath)
 
 int main(int argc, char * * argv)
 {
-    // I *think* it's safe to assume that a path from a symbolic link
-    // should safely fit within the PATH_MAX system limit. Though I'm
-    // not positive it's safe...
-    char selfPath[PATH_MAX];
-    int selfPathSize = readlink("/proc/self/exe", selfPath, sizeof(selfPath));
+    int selfExe = open("/proc/self/exe", O_PATH | O_CLOEXEC | O_NOFOLLOW);
+    assert(selfExe != -1);
```

Comment:
One of my todos after adding setcap wrapper support that got lost in the cracks was to improve error messaging in the wrapper program. What do you think about checking the return and printing the error code to the user? Or perhaps we can do that if the debug env var is set? I think it would be useful if we did that.

Don't consider this a blocking comment, if you agree, we can address it in a refactoring-pass patch.

Reply:
I do agree, yes. No need for a debug env var I think, the error path is allowed to be slow :)

Reply:
@shlevy we already support a `WRAPPER_DEBUG` env var in this program code, so we could use that if you felt that was a slightly nicer way to surface these wrapper-specific errors to the user.

For example, you can turn it on like this:

```
$ WRAPPER_DEBUG=1 ping google.com
cap_setpcap in set, skipping it
raised cap_net_raw+p into the Ambient capability set
PING google.com(iad23s42-in-x200e.1e100.net (2607:f8b0:4004:80c::200e)) 56 data bytes
64 bytes from iad23s42-in-x200e.1e100.net (2607:f8b0:4004:80c::200e): icmp_seq=1 ttl=51 time=40.1 ms
64 bytes from iad23s42-in-x200e.1e100.net (2607:f8b0:4004:80c::200e): icmp_seq=2 ttl=51 time=39.9 ms
```

Reply:
I think `WRAPPER_DEBUG` makes sense for debug info, but on the error path I don't think it's necessary. Especially since it's often too late to turn on `WRAPPER_DEBUG` when you hit an issue :smile:

Reply:
Yeah, that is a good point.